src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,152 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+        /// <paramref name="stringCallback"/> controls whether to continue the visit by returning true or false:
+        /// if true, the visit will continue. If false, the walk will be cut off.
+        /// </summary>
+        public static bool VisitBinaryOperatorInterpolatedString<TInterpolatedStringType, TArg>(
+            this BoundBinaryOperator binary,
+            TArg arg,
+            Func<TInterpolatedStringType, TArg, bool> stringCallback,
+            Action<BoundBinaryOperator, TArg>? binaryOperatorCallback = null)
+            where TInterpolatedStringType : BoundInterpolatedStringBase
+        {
+            var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+
+            pushLeftNodes(binary, stack);
+
+            while (stack.TryPop(out BoundBinaryOperator? current))
+            {
+                switch (current.Left)
+                {
+                    case BoundBinaryOperator:
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!stringCallback(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Left.Kind);
+                }
+
+                binaryOperatorCallback?.Invoke(current, arg);
+
+                switch (current.Right)
+                {
+                    case BoundBinaryOperator rightOperator:
+                        pushLeftNodes(rightOperator, stack);
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!stringCallback(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Right.Kind);
+                }
+            }
+
+            Debug.Assert(stack.Count == 0);
+            stack.Free();
+            return true;
+
+            static void pushLeftNodes(BoundBinaryOperator binary, ArrayBuilder<BoundBinaryOperator> stack)
+            {
+                Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString) || binary.IsUnconvertedInterpolatedStringAddition);
+                BoundBinaryOperator? current = binary;
+                while (current != null)
+                {
+                    stack.Push(current);
+                    current = current.Left as BoundBinaryOperator;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Rewrites a BoundBinaryOperator composed of interpolated strings (either converted or unconverted) iteratively, without
+        /// recursion on the left side of the tree. Nodes of the tree are rewritten in a depth-first post-order fashion.
```

Comment:
>depth-first post-order [](http://example.com/codeflow?start=89&length=22)

I think it would be clearer to specify the order explicitly. There are three nodes (left, right, parent), in what order are they supposed to be visited?